auth/oidc.go

```diff
@@ -439,3 +533,180 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
 	uri.RawQuery = rawQuery.Encode()
 	return uri.String(), nil
 }
+
+// StartDiscoverySync runs the provider metadata discovery sync task in background.
+func (op *OIDCProvider) StartDiscoverySync(discoveryURL string) error {
+	op.stopDiscoverySync = make(chan struct{})
+	duration, err := op.runDiscoverySync(discoveryURL)
+	if err != nil {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-time.After(duration):
+				duration, err = op.runDiscoverySync(discoveryURL)
+				if err != nil {
+					base.Warnf("OpenID Connect provider discovery sync ends up in error: %v, next retry in %v", err, duration)
+				}
+			case <-op.stopDiscoverySync:
+				base.Debugf(base.KeyAll, "Terminating OpenID Connect provider discovery sync")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// runDiscoverySync polls the provider metadata from discovery endpoint and refresh the ID token verifier.
+func (op *OIDCProvider) runDiscoverySync(discoveryURL string) (time.Duration, error) {
+	base.Debugf(base.KeyAuth, "Syncing provider metadata from: %s", base.UD(discoveryURL))
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("unsuccessful response: %v", err)
+	}
+
+	var expiresAt time.Time
+	var ttl time.Duration
+	var ok bool
+	ttl, ok, err = cacheable(resp.Header)
+	if err != nil {
+		return 0, err
+	} else if ok && ttl > 0 {
+		expiresAt = time.Now().UTC().Add(ttl)
+	}
+
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, err
+	}
+	hash := sha1.New()
+	_, _ = hash.Write(bodyBytes)
+	metadataChecksum := hex.EncodeToString(hash.Sum(nil))
+
+	if op.metadataChecksum == metadataChecksum {
+		base.Debugf(base.KeyAuth, "No change in discovery config detected at this time, next sync will be at %v", expiresAt)
+		return op.nextSyncAfter(expiresAt), nil
+	}
+	op.metadataChecksum = metadataChecksum
+
+	var metadata ProviderMetadata
+	if err := base.JSONUnmarshal(bodyBytes, &metadata); err != nil {
+		base.Debugf(base.KeyAuth, "Error parsing body during discovery sync: %v", err)
+		return 0, err
+	}
+
+	if metadata.Issuer != op.Issuer {
+		return 0, fmt.Errorf("oidc: issuer did not match the issuer returned by provider, expected %q got %q", op.Issuer, metadata.Issuer)
+	}
+	verifier := op.GenerateVerifier(&metadata, context.Background())
```

Comment:
If we're using the default discovery endpoint, we previously were trying to leverage the OIDC library for verifier generation (in DiscoverConfig).  The handling here means that the first time we identify that the content on that endpoint changes, we're switching to the non-library handling.  That's not ideal.

I think we have two options:
1. Abandon using the library for config retrieval, even for default provider endpoints.  This has the drawback of having us clone library functionality, and make it easier to miss library fixes in this handling.
2. Have runDiscoverSync *only* identify whether the content on the endpoint has changed, and trigger DiscoverSync when it detects a change.  This has the drawback of an extra round-trip to the endpoint to load the config in the default providerURL case.

Both of these approaches would avoid the duplicated code we've got in runDiscoverSync and DiscoverConfig today.

Can you review and decide which of these (or a third option, if you see one) we should take?

Reply:
Makes sense, selected the second approach to avoid the duplicated code we've got in runDiscoverSync and DiscoverConfig today.